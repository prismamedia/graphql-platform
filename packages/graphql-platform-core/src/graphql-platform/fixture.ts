import { Entries, fromEntries } from '@prismamedia/graphql-platform-utils';
import { DepGraph } from 'dependency-graph';
import { OperationDefinitionNode, print } from 'graphql';
import { Logger } from 'winston';
import { GraphQLPlatform, GraphQLRequest } from '../graphql-platform';
import { CreateOneDataInputValue } from './operation';
import { AnyResource, Relation, SerializedFieldValue } from './resource';
import { TypeKind, WhereUniqueInputValue } from './type';

export class FixtureGraph extends DepGraph<Fixture> {}

export type FixtureFieldData = SerializedFieldValue;

export type FixtureRelationData = null | Fixture['name'];

export type FixtureComponentData = FixtureFieldData | FixtureRelationData;

export type FixtureData = {
  [componentName: string]: FixtureComponentData;
};

export type FixtureDataMap = {
  [fixtureName: string]: FixtureData;
};

export class Fixture {
  readonly logger?: Logger;
  protected id?: WhereUniqueInputValue;

  public constructor(
    readonly name: string,
    readonly resource: AnyResource,
    readonly data: FixtureData,
    protected graph: FixtureGraph,
    protected gp: GraphQLPlatform,
  ) {
    this.logger = gp.getLogger();
  }

  public toString(): string {
    return `${this.resource}.${this.name}`;
  }

  public getRelatedFixture(relation: Relation): Fixture | null {
    this.resource.getRelationMap().assert(relation);

    const relationValue = this.data[relation.name];

    if (relationValue == null) {
      if (relation.isRequired()) {
        throw new Error(`In the fixture "${this.name}", the relation "${relation}" can't be null.`);
      }
    } else if (!(typeof relationValue === 'string' && this.graph.hasNode(relationValue))) {
      throw new Error(
        `In the fixture "${this.name}", the relation "${relation}" has to reference an "${relation.getTo()}" fixture.`,
      );
    }

    return relationValue ? this.graph.getNodeData(relationValue) : null;
  }

  public getCreateMutationAST(): OperationDefinitionNode {
    const operationDefinitionNode: OperationDefinitionNode = {
      kind: 'OperationDefinition',
      operation: 'mutation',
      variableDefinitions: [
        {
          kind: 'VariableDefinition',
          variable: {
            kind: 'Variable',
            name: {
              kind: 'Name',
              value: 'data',
            },
          },
          type: {
            kind: 'NonNullType',
            type: {
              kind: 'NamedType',
              name: {
                kind: 'Name',
                value: this.resource.getMutation('CreateOne').getDataType().name,
              },
            },
          },
        },
      ],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            name: {
              kind: 'Name',
              value: this.resource.getMutation('CreateOne').name,
            },
            alias: {
              kind: 'Name',
              value: 'id',
            },
            arguments: [
              {
                kind: 'Argument',
                name: {
                  kind: 'Name',
                  value: 'data',
                },
                value: {
                  kind: 'Variable',
                  name: {
                    kind: 'Name',
                    value: 'data',
                  },
                },
              },
            ],
            selectionSet: this.resource
              .getFirstPublicUnique()
              .getSelectionNode(TypeKind.Output)
              .toAST(),
          },
        ],
      },
    };

    return operationDefinitionNode;
  }

  public getCreateMutationSource(): string {
    return print(this.getCreateMutationAST());
  }

  public getCreateMutationVariables(): GraphQLRequest['variableValues'] {
    return {
      data: fromEntries([
        // Fields
        ...[...this.resource.getFieldSet()].map(field =>
          !field.isFullyManaged() ? [field.name, this.data[field.name] as FixtureFieldData] : undefined,
        ),
        // Relations
        ...[...this.resource.getRelationSet()].map(relation => {
          if (!relation.isFullyManaged()) {
            const relatedFixture = this.getRelatedFixture(relation);
            if (relatedFixture) {
              return [relation.name, { connect: relatedFixture.assertId() }];
            }
          }
        }),
      ] as Entries<CreateOneDataInputValue>),
    };
  }

  public getCreateMutation(): GraphQLRequest {
    return {
      source: this.getCreateMutationSource(),
      variableValues: this.getCreateMutationVariables(),
    };
  }

  public async load(contextValue?: GraphQLRequest['contextValue']): Promise<void> {
    try {
      const { id } = await this.gp.execute<{ id: WhereUniqueInputValue }>({
        ...this.getCreateMutation(),
        contextValue,
      });

      this.id = this.resource.getInputType('WhereUnique').assert(id);
    } catch (error) {
      this.logger && this.logger.error(error);

      throw error;
    }
  }

  public assertId(): WhereUniqueInputValue {
    if (!this.id) {
      throw new Error(`The fixture "${this.name}" is not loaded yet.`);
    }

    return this.id;
  }
}
