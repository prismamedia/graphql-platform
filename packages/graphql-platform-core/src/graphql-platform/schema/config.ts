import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
