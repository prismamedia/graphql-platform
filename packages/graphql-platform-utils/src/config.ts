export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
