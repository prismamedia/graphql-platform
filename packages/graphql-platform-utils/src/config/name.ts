import * as graphql from 'graphql';
import { UnexpectedValueError } from '../error.js';
import type { Path } from '../path.js';

/**
 * A string compliant with the GraphQL specification
 *
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export type Name = string;

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function assertName(
  maybeName: unknown,
  path: Path,
): asserts maybeName is Name {
  if (typeof maybeName !== 'string' || !maybeName) {
    throw new UnexpectedValueError(`a non-empty string`, maybeName, {
      path,
    });
  }

  try {
    graphql.assertName(maybeName);
  } catch (error) {
    throw new UnexpectedValueError(
      'to be valid against the GraphQL "Names" specification (@see: https://spec.graphql.org/draft/#sec-Names)',
      maybeName,
      { cause: error, path },
    );
  }
}

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function ensureName(maybeName: unknown, path: Path): Name {
  assertName(maybeName, path);

  return maybeName;
}
