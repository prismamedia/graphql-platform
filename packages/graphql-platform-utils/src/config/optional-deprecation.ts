import { UnexpectedValueError } from '../error.js';
import { isNil, Nillable } from '../nil.js';
import type { Path } from '../path.js';

export type OptionalDeprecation = Nillable<boolean | string>;

export function assertOptionalDeprecation(
  maybeOptionalDeprecation: unknown,
  path: Path,
): asserts maybeOptionalDeprecation is OptionalDeprecation {
  if (
    !isNil(maybeOptionalDeprecation) &&
    typeof maybeOptionalDeprecation !== 'boolean' &&
    typeof maybeOptionalDeprecation !== 'string'
  ) {
    throw new UnexpectedValueError(
      `a boolean or a string`,
      maybeOptionalDeprecation,
      { path },
    );
  }
}

export function ensureOptionalDeprecation(
  maybeOptionalDeprecation: unknown,
  path: Path,
): OptionalDeprecation {
  assertOptionalDeprecation(maybeOptionalDeprecation, path);

  return maybeOptionalDeprecation;
}

export function getOptionalDeprecation(
  maybeOptionalDeprecation: OptionalDeprecation,
  defaultValue: string,
  path: Path,
): string | undefined {
  const optionalDeprecation = ensureOptionalDeprecation(
    maybeOptionalDeprecation,
    path,
  );

  return optionalDeprecation === true
    ? defaultValue
    : optionalDeprecation || undefined;
}
