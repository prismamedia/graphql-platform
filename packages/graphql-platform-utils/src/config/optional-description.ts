import { UnexpectedConfigError } from '../error.js';
import type { Path } from '../path.js';

export type OptionalDescription = string | null | undefined;

export function assertOptionalDescription(
  maybeOptionalDescription: unknown,
  path: Path,
): asserts maybeOptionalDescription is OptionalDescription {
  if (
    maybeOptionalDescription != null &&
    typeof maybeOptionalDescription !== 'string'
  ) {
    throw new UnexpectedConfigError(`a string`, maybeOptionalDescription, {
      path,
    });
  }
}

export function ensureOptionalDescription(
  maybeOptionalDescription: unknown,
  path: Path,
): OptionalDescription {
  assertOptionalDescription(maybeOptionalDescription, path);

  return maybeOptionalDescription;
}

export const getOptionalDescription = (
  maybeOptionalDescription: OptionalDescription,
  path: Path,
): string | undefined =>
  ensureOptionalDescription(maybeOptionalDescription, path) || undefined;
