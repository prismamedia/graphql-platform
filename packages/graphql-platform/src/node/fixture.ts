import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import type { ConnectorInterface } from '../connector-interface.js';
import type { Node, NodeValue } from '../node.js';
import type { Seeding } from '../seeding.js';
import type { UniqueConstraintValue } from './definition.js';
import { EdgeCreationInputAction } from './type/input/creation/field.js';

export type NodeFixtureReference = string;

export type NodeFixtureData = utils.PlainObject;

export class NodeFixture<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
> {
  public readonly dependencies: ReadonlySet<NodeFixtureReference>;

  public constructor(
    public readonly seeding: Seeding<TRequestContext, TConnector>,
    public readonly node: Node<TRequestContext, TConnector>,
    public readonly reference: NodeFixtureReference,
    public readonly data: NodeFixtureData,
  ) {
    if (typeof reference !== 'string') {
      throw new utils.UnexpectedValueError('a string', reference);
    }

    if (!utils.isPlainObject(data)) {
      throw new utils.UnexpectedValueError('a plain-object', data);
    }

    this.dependencies = new Set(
      Array.from(this.node.edgesByName.values()).reduce<NodeFixtureReference[]>(
        (dependencies, edge) => {
          const maybeEdgeReference = data[edge.name];
          if (maybeEdgeReference != null) {
            dependencies.push(maybeEdgeReference);
          }

          return dependencies;
        },
        [],
      ),
    );
  }

  @Memoize()
  public get dependents(): ReadonlySet<NodeFixtureReference> {
    return new Set(
      Array.from(this.seeding.fixturesByReference.values()).reduce<
        NodeFixtureReference[]
      >((dependents, fixture) => {
        if (fixture.dependencies.has(this.reference)) {
          dependents.push(fixture.reference);
        }

        return dependents;
      }, []),
    );
  }

  @Memoize((context: TRequestContext) => context)
  public async load(context: TRequestContext): Promise<NodeValue> {
    const data = this.node.creationInputType.parseValue(
      Object.fromEntries(
        await Promise.all(
          Object.entries(this.data).map(
            async ([componentName, componentValue]) => [
              componentName,
              this.node.edgesByName.has(componentName)
                ? {
                    [EdgeCreationInputAction.CONNECT]:
                      await this.seeding.fixturesByReference
                        .get(componentValue)!
                        .getIdentifier(context),
                  }
                : componentValue,
            ],
          ),
        ),
      ),
      utils.addPath(undefined, this.reference),
    ) as any;

    return this.node
      .getMutationByKey('create-one')
      .execute({ data, selection: this.node.selection }, context) as any;
  }

  @Memoize((context: TRequestContext) => context)
  public async getIdentifier(
    context: TRequestContext,
  ): Promise<UniqueConstraintValue> {
    const nodeValue = await this.load(context);

    return this.node.identifier.parseValue(nodeValue);
  }
}
