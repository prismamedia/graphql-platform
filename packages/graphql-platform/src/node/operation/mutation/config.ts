import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
