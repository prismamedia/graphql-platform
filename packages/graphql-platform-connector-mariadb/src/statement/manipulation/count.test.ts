import { afterAll, beforeAll, describe, expect, it } from '@jest/globals';
import {
  ArticleStatus,
  myAdminContext,
  myUserContext,
} from '@prismamedia/graphql-platform/__tests__/config.js';
import { fixtures } from '@prismamedia/graphql-platform/__tests__/fixture.js';
import { format } from '@sqltools/formatter';
import { createMyGP, type MyGP } from '../../__tests__/config.js';

describe('Count statement', () => {
  let gp: MyGP;
  const executedStatements: string[] = [];

  beforeAll(async () => {
    gp = createMyGP(`connector_mariadb_count_statement`);
    gp.connector.on('executed-statement', ({ statement }) =>
      executedStatements.push(format(statement.sql).replaceAll('<= >', '<=>')),
    );

    await gp.connector.setup();
    await gp.seed(fixtures, myAdminContext);
  });

  afterAll(async () => {
    await gp.connector.teardown();
  });

  it.each([
    ['Article', myUserContext, undefined],
    ['Article', myAdminContext, undefined],
    [
      'Article',
      myAdminContext,
      {
        where: {
          status: ArticleStatus.PUBLISHED,
          body_contains: 'my searched content',
          tagCount_gte: 5,
        },
      },
    ],
  ])('generates statements', async (nodeName, context, args) => {
    executedStatements.length = 0;

    await expect(
      gp.getNodeByName(nodeName).getQueryByKey('count').execute(context, args),
    ).resolves.toMatchSnapshot('result');

    expect(executedStatements).toMatchSnapshot('statements');
  });
});
