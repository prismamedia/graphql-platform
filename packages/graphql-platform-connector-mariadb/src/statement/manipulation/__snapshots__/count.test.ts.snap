// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Count statement generates statements: result 1`] = `4`;

exports[`Count statement generates statements: result 2`] = `6`;

exports[`Count statement generates statements: result 3`] = `0`;

exports[`Count statement generates statements: statements 1`] = `
[
  "SELECT COUNT(*) AS COUNT
FROM \`articles\`
WHERE \`articles\`.\`status\` <=> 'published'",
]
`;

exports[`Count statement generates statements: statements 2`] = `
[
  "SELECT COUNT(*) AS COUNT
FROM \`articles\`",
]
`;

exports[`Count statement generates statements: statements 3`] = `
[
  "SELECT COUNT(*) AS COUNT
FROM \`articles\`
WHERE (
    \`articles\`.\`status\` <=> 'published'
    AND JSON_SEARCH(
      \`articles\`.\`body\`,
      'one',
      '%my searched content%',
      NULL,
      '$.blocks[*].text'
    ) IS NOT NULL
    AND (
      SELECT COUNT(*)
      FROM \`article_tags\` AS \`articles>tags\`
      WHERE \`articles\`.\`private_id\` <=> \`articles>tags\`.\`article_private_id\`
    ) > 4
  )",
]
`;
